### PR TITLE
drivers:dac: added max538x family support

### PR DESCRIPTION
**Pull request by jsanbuen**
added initial driver implementation for max538x family.

For more info: https://www.analog.com/media/en/technical-documentation/data-sheets/MAX5380-MAX5382.pdf

Signed-off-by: Jose Ramon San Buenaventura <jose.sanbuenaventura@analog.com>

### CODE REVIEW

**Comment by buha** [comment]
also drop the merge commit, it should be just 1 commit with your changes

File: drivers/dac/max538x/max538x.c [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+
+/***************************************************************************//**
+ * @brief I2C transaction
+ * @param dev - The device structure.
+ * @param data - pointer to the data to write on the bus
+ * @return 0 in case of success, error code otherwise
+*******************************************************************************/
+int max538x_i2c_transaction(struct max538x_dev *dev, uint8_t *data)
+{
+	uint8_t data_buffer[1];
```
Comment (amiclaus) [inline drivers/dac/max538x/max538x.c:109]: is this extra variable needed?

File: drivers/dac/max538x/max538x.c [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+	struct max538x_dev *dev;
+	int32_t status;
+
+	dev = (struct max538x_dev *)malloc(sizeof(*dev));
+	if (!dev)
+		return -ENOMEM;
+
+	if( (init_param.max538x_vdd == NULL || init_param.max538x_vdd == 0)
+	    && dev->active_device > MAX5381K)
+		return -EINVAL;
```
Comment (amiclaus) [inline drivers/dac/max538x/max538x.c:142]: we should free dev in this case

File: drivers/dac/max538x/max538x.c [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+
+	if(dev->active_device <= MAX5381K)
+		dev->max538x_vref = chip_info[init_param.active_device].vfactor;
+	else
+		dev->max538x_vref = chip_info[init_param.active_device].vfactor *
+				    init_param.max538x_vdd;
+
+	status = no_os_i2c_init(&dev->i2c_desc, &init_param.i2c_init);
+
+	if (status < 0)
```
Comment (amiclaus) [inline drivers/dac/max538x/max538x.c:155]: nit: `if (status)`

File: drivers/dac/max538x/max538x.h [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+struct max538x_chip_info {
+	uint8_t vfactor;
+	uint8_t addr;
+};
+
+struct max538x_dev {
+	/* I2C */
+	struct no_os_i2c_desc	*i2c_desc;
+
```
Comment (amiclaus) [inline drivers/dac/max538x/max538x.h:81]: remove all these extra empty lines

**Review by amiclaus — COMMENTED** [review]
looks good overall, some minor comments on my side.

File: drivers/dac/max538x/max538x.c [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+		     struct max538x_init_param init_param)
+{
+	struct max538x_dev *dev;
+	int32_t status;
+
+	dev = (struct max538x_dev *)malloc(sizeof(*dev));
+	if (!dev)
+		return -ENOMEM;
+
+	if( (init_param.max538x_vdd == NULL || init_param.max538x_vdd == 0)
```
Comment (amiclaus) [inline drivers/dac/max538x/max538x.c:140]: `if ((`

File: drivers/dac/max538x/max538x.c [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+	    && dev->active_device > MAX5381K)
+		return -EINVAL;
+
+	dev->active_device = init_param.active_device;
+	init_param.i2c_init.slave_address = chip_info[init_param.active_device].addr;
+
+	if(dev->active_device <= MAX5381K)
```
Comment (amiclaus) [inline drivers/dac/max538x/max538x.c:147]: `if (`

File: drivers/dac/max538x/max538x.c [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+ *
+ * @param dev - The device structure.
+ *
+ * @return - 0 if no error, else return error code of I2C transaction
+*******************************************************************************/
+int max538x_set_voutput(struct max538x_dev *dev, float vout)
+{
+	uint8_t vout_code;
+
+	if(vout < 0 || vout > dev->max538x_vref)
```
Comment (amiclaus) [inline drivers/dac/max538x/max538x.c:196]: `if (`

File: drivers/dac/max538x/max538x.h [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+/******************************************************************************/
+#include "no_os_i2c.h"
+#include "no_os_util.h"
+#include <stdbool.h>
+
+/******************************************************************************/
+/************************** max538x Definitions *******************************/
+/******************************************************************************/
+/* MAX5380 default resolution */
+#define MAX538x_resolution 		0xFF
```
Comment (amiclaus) [inline drivers/dac/max538x/max538x.h:53]: use only uppercase for macro definition

**Comment by jsanbuen** [comment]
Changelog/s:
- corrected format based on comments i.e., if ()..
- added "free(dev)" in init condition for inappropriate initial settings.
- changed macro definition for device resolution to all uppercase.
- removed extra variable in max538x_transaction function based on reviewer comment.

**Comment by jsanbuen** [comment]
Changelog/s:
-added errno and other libraries to max538x.c and max538x.h files

**Review by buha — COMMENTED** [review]
i forgot to publish my review yesterday, i think some of the comments may be outdated

**Comment by jsanbuen** [comment]
Changelog/s:
- resolved typo on 5380x (changed to 538x).
- updated dates to 2023(c)
- dropped max538x_i2c_transaction function to use no_os_i2c_write directly.
- changed from malloc to calloc.
- removed whitespaces.

**Comment by jsanbuen** [comment]
Changelog/s:
-changed malloc to calloc as recommended on reviewer comment.
-removed whitespaces.
-changed the max5380x to max538x in the comment / header.
-update the date to 2023(c).
-removed max538x_i2c_transaction to use the no_os_i2c_write function directly.

File: drivers/dac/max538x/max538x.c [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+
+	status = no_os_i2c_init(&dev->i2c_desc, &init_param.i2c_init);
+	if (status)
+		goto max538x_init_error;
+
+	*device = dev;
+
+	return 0;
+
+max538x_init_error:
```
Comment (buha) [inline drivers/dac/max538x/max538x.c:149]: you can use simple labels like:
`error:`

no need to prefix them with anything, their scope is a function so there won't be name clashes

File: drivers/dac/max538x/max538x.c [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+	max538x_remove(dev);
```
Comment (buha) [inline drivers/dac/max538x/max538x.c:150]: best here would be to not use max538x_remove() but manually perform:
```
error_1:
    free(dev);
error:
    no_os_i2c_remove(dev->i2c_desc);
```

And jump to appropriate label with goto.

File: drivers/dac/max538x/max538x.c [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+	struct max538x_dev *dev;
+	int32_t status;
+
+	dev = (struct max538x_dev *)calloc(0, sizeof(*dev));
+	if (!dev)
+		return -ENOMEM;
+
+	if ((init_param.max538x_vdd == NULL || init_param.max538x_vdd == 0)
+	    && dev->active_device > MAX5381K) {
+		free(dev);
```
Comment (buha) [inline drivers/dac/max538x/max538x.c:128]: (see my other comment below)
here, you could `goto error;`

File: drivers/dac/max538x/max538x.c [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+
+	if (dev->active_device <= MAX5381K)
+		dev->max538x_vref = chip_info[init_param.active_device].vfactor;
+	else
+		dev->max538x_vref = chip_info[init_param.active_device].vfactor *
+				    init_param.max538x_vdd;
+
+	status = no_os_i2c_init(&dev->i2c_desc, &init_param.i2c_init);
+	if (status)
+		goto max538x_init_error;
```
Comment (buha) [inline drivers/dac/max538x/max538x.c:143]: (see my other comment below)
and here, you could `goto error_1;`

**Comment by jsanbuen** [comment]
Changelog/s:
-used simplified error labels with goto for error handling in init function.
-changed calloc(0, ....) to calloc(1, ....) based on observed from driver code w/ calloc (ADXL355).

**Comment by jsanbuen** [comment]
Changelog/s:
-in init funciton, parameter checking for EINVAL return is placed ahead of other checking.
-corrected goto and label use for when there's error in no_os_i2c_init function based on reviewer comments.

File: drivers/dac/max538x/max538x.c [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+ * @param init_param - The structure that contains the device initial
+ * 		       parameters.
+ *
+ * @return status - The result of the initialization procedure.
+ *                  Example: -1 - I2C peripheral was not initialized or the
+ *                                device is not present.
+ *                            0 - I2C peripheral was initialized and the
+ *                                device is present.
+*******************************************************************************/
+int32_t max538x_init(struct max538x_dev **device,
```
Comment (amiclaus) [inline drivers/dac/max538x/max538x.c:116]: nit: we switched to using standard `int` for the function types. please update them accordingly (like you did for `int max538x_set_voutput(struct max538x_dev *dev, float vout)`)

**Comment by jsanbuen** [comment]
Changelog/s
-changed function types to int from int32_t for init and remove functions.

**Review by buha — APPROVED** [review]

File: drivers/dac/max538x/max538x.c [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+}
+
+/***************************************************************************//**
+ * @brief Free the resources allocated by max538x_init().
+ * @param dev - The device structure.
+ * @return ret - The result of the remove procedure.
+*******************************************************************************/
+int max538x_remove(struct max538x_dev *dev)
+{
+	int32_t ret;
```
Comment (amiclaus) [inline drivers/dac/max538x/max538x.c:159]: `ret` type should match the function type.

File: drivers/dac/max538x/max538x.c [hunk 15]
```diff
@@ -0,0 +1,185 @@
...
+ *                  Example: -1 - I2C peripheral was not initialized or the
+ *                                device is not present.
+ *                            0 - I2C peripheral was initialized and the
+ *                                device is present.
+*******************************************************************************/
+int max538x_init(struct max538x_dev **device,
+		 struct max538x_init_param init_param)
+{
+	struct max538x_dev *dev;
+	int32_t status;
```
Comment (amiclaus) [inline drivers/dac/max538x/max538x.c:120]: `status` type should match the function type. Also consider using throughout the code only one naming: either `status`, either `ret`

**Comment by buha** [comment]
in order to merge, @jsanbuen please fix the cppcheck and remaining comments from @amiclaus

**Comment by jsanbuen** [comment]
Changelog/s:
- changed int32_t types to int.
- consistent use of variable ret for all other functions.

**Comment by jsanbuen** [comment]
Changelog/s:
-in init function, corrected the initial EINVAL check by changing dev->active_device to init_param.active device. Checked cppcheck in pipeline and this part was initially source of error in cppcheck.

**Review by amiclaus — APPROVED** [review]

**Pull request merged**